src/client/Client.cc

```diff
@@ -3091,7 +3091,19 @@ void Client::handle_mds_map(const MConstRef<MMDSMap>& m)
 	continue;
       }
       if (newstate >= MDSMap::STATE_ACTIVE) {
-	if (oldstate < MDSMap::STATE_ACTIVE) {
+	if (oldstate <= MDSMap::STATE_ACTIVE) {
+	  ceph_assert(newstate != oldstate);
+
+          // flush the delayed caps in case this MDS is stopping
+          for (auto p = delayed_list.begin(); p != delayed_list.end(); ) {
+            Inode *in = *p;
+            ++p;
+            if (!mount_aborted && in->auth_cap->session == session.get()) {
+              in->delay_cap_item.remove_myself();
+              check_caps(in, CHECK_CAPS_NODELAY);
+	    }
+          }
```

Comment:
I don't understand at all the purpose of flushing the caps here. We can never guarantee they will get flushed before the MDS stops, and something else should take over for it, so why bother doing something that adds load to a daemon right when it's trying to shed load?

Reply:
It's going to do this eventually anyway in the tick thread. We might as well flush as soon as the MDS is stopping and then try to stop communicating with it.